Apps/Src/SendTritium.c

```diff
@@ -0,0 +1,426 @@
+/**
+ * @copyright Copyright (c) 2022 UT Longhorn Racing Solar
+ * 
+ * @file SendTritium.c
+ * @brief Function implementations for the SendTritium application.
+ * 
+ * This contains functions relevant to updating the velocity and current setpoitns
+ * of the Tritium motor controller. The implementation includes a normal current
+ * controlled mode, a one-pedal driving mode, cruise control, and regenerative braking.
+ * The logic is determined through a finite state machine implementation.
+ * 
+ * @author Nathaniel Delgado (NathanielDelgado)
+ * @author Diya Rajon (diyarajon)
+ * @author Ishan Deshpande (IshDeshpa)
+ */
+
+#include "Pedals.h"
+#include "ReadCarCAN.h"
+#include "Minions.h"
+#include "SendTritium.h"
+#include "ReadTritium.h"
+#include "CANbus.h"
+
+
+// Macros
+#define MAX_VELOCITY 50.0f // m/s    if rmp: 20000.0f // rpm
+#define NORMAL_BRAKE_THRESHOLD 5  // percent
+#define ONEPEDAL_BRAKE_THRESHOLD 20  // percent
+#define NEUTRAL_THRESHOLD 25    // percent
+#define DEBOUNCE_PERIOD 2 // in 100 ms
+#define VELOCITY_THRESHOLD 20.0f // m/s
+
+#define MOTOR_MSG_PERIOD 100
+#define FSM_PERIOD 20
+#define MOTOR_MSG_COUNTER_THRESHOLD (MOTOR_MSG_PERIOD)/(FSM_PERIOD)
+
+#ifdef __TEST_SENDTRITIUM
+#define SCOPE
+#else
+#define SCOPE static
+static int MotorCounter = 0;
+#endif
+
+// Inputs
+SCOPE bool cruiseEnable = false;
+SCOPE bool cruiseSet = false;
+SCOPE bool regenToggle = false;
+SCOPE uint8_t brakePedalPercent = 0;
+SCOPE uint8_t accelPedalPercent = 0;
+SCOPE bool reverseSwitch = false;
+SCOPE bool forwardSwitch = false;
+
+// Outputs
+SCOPE float currentSetpoint = 0;
+SCOPE float velocitySetpoint = 0;
+SCOPE float cruiseVelSetpoint = 0;
+static bool brakelightState = false;
+
+#ifndef __TEST_SENDTRITIUM
+// Debouncing counters
+static uint8_t regenCounter = 0;
+static uint8_t cruiseEnableCounter = 0;
+static uint8_t cruiseSetCounter = 0;
+
+// Button states
+static bool regenButton = false;
+static bool regenPrevious = false;
+
+static bool cruiseEnableButton = false;
+static bool cruiseEnablePrevious = false;
+
+#else
+float velocityObserved = 0;
```

Comment:
If this is being used or changed in the test file, it can have the SCOPE macro in front of it to conditionally expose/not expose the variable based on whether we're testing or not.